particle-iot/device-os
# System power management inprovements

### Problem
The battery state is not reported consistently.
### Solution
1. When charging is enabled (through the `HAL_POWER_CHARGE_STATE_DISABLE` flag), we debounce the CHARGING/CHARGED/CHARGING/CHARGED events.
2. When charging is disabled, we use the `VCELL` of the fuel gauge to detect the battery state.
### Steps to Test
1. Flashing the attached application to the platforms with PMIC and fuel gauge installed
2. Attach/detach batteries with different capacity and observe the reported battery state.
3. Plug/unplug the power source and observe the reported battery state, as well as the reported power source
### Example App

```c
#include "Particle.h"

SYSTEM_THREAD(ENABLED);
SYSTEM_MODE(SEMI_AUTOMATIC);

Serial1LogHandler logHandler(115200, LOG_LEVEL_ALL);

static system_tick_t loopTick = 0;

constexpr char const* batteryStates[] = {
  "unknown", "not charging", "charging",
  "charged", "discharging", "fault", "disconnected"
};
constexpr char const* powerSources[] = {
  "unknown", "vin", "usb host", "usb adapter",
  "usb otg", "battery"
};

void setup() {
  Serial1.begin(115200);
  Serial1.printlnf("Application started.");

  System.on(battery_state, [](system_event_t event, int data) -> void {
    Serial1.printlnf("battery_state: %s", batteryStates[std::max(0, data)]);
  });

  System.on(power_source, [](system_event_t event, int data) -> void {
    Serial1.printlnf("power_source: %s", powerSources[std::max(0, data)]);
  });

  System.on(low_battery, [](system_event_t event, int data) -> void {
    Serial1.printlnf("low_battery: %0.2f", System.batteryCharge());
  });

  Particle.connect();
}

void loop() {
  if (Serial1.available()) {
    char c = (char)Serial1.read();
    // Commands are case sensitive
    switch (c) {
      case 'g': {
        auto cfg = System.getPowerConfiguration();
        Serial1.printlnf("cmd, getconfig, %d, %d, %d, %d, %u, %u, %u, %u",
          cfg.isFeatureSet(SystemPowerFeature::PMIC_DETECTION) ? 1 : 0,
          cfg.isFeatureSet(SystemPowerFeature::USE_VIN_SETTINGS_WITH_USB_HOST) ? 1 : 0,
          cfg.isFeatureSet(SystemPowerFeature::DISABLE) ? 1 : 0,
          cfg.isFeatureSet(SystemPowerFeature::DISABLE_CHARGING) ? 1 : 0,
          cfg.powerSourceMinVoltage(),
          cfg.powerSourceMaxCurrent(),
          cfg.batteryChargeVoltage(),
          cfg.batteryChargeCurrent()
        );
        break;
      }
      case '0': { // Disable charging
        int ret = System.setPowerConfiguration(System.getPowerConfiguration().feature(SystemPowerFeature::DISABLE_CHARGING));
        Serial1.printlnf("Disable charging, %d", ret);
        break;
      }
      case '1': { // Enable charging
        int ret = System.setPowerConfiguration(System.getPowerConfiguration().clearFeature(SystemPowerFeature::DISABLE_CHARGING));
        Serial1.printlnf("Enable charging, %d", ret);
        break;
      }
      default: {
        break;
      }
    }
  }

  auto now = millis();
  if ((now - loopTick) >= 2000) {
    loopTick = now;

    FuelGauge fuel;
    int powerSource = System.powerSource();
    int batteryState = System.batteryState();
    float batterySoc = System.batteryCharge();
    Serial1.printlnf("power: %s, batt: %s, SoC: %0.2f, Vol: %0.2f",
      powerSources[std::max(0, powerSource)], batteryStates[std::max(0, batteryState)],
      batterySoc, fuel.getVCell());
  }
}
```
### Completeness
